nginx/modsecurity/ngx_http_modsecurity.c

```diff
@@ -808,13 +808,14 @@ ngx_http_modsecurity_detection_task_offload(ngx_http_request_t *r)
     ngx_http_modsecurity_thread_ctx_t *thread_ctx = task->ctx;
     thread_ctx->r = r;
 
-    task->handler = ngx_http_modsecurity_detection_thread_func;
+    task->handler = ngx_http_modsecurity_detection_run_processing;
     task->event.handler = ngx_http_modsecurity_detection_thread_completion;
     task->event.data = r;
 
     ngx_thread_pool_t* thread_pool = ngx_thread_pool_get((ngx_cycle_t *)ngx_cycle, &thread_pool_name);
     if (ngx_thread_task_post(thread_pool, task) != NGX_OK) {
-        return NGX_ERROR;
+        ngx_http_modsecurity_detection_run_processing(task->ctx, r->connection->log);
```

Comment:
so this will run in the main thread?

Reply:
Yes, in case if we fail to post the task to the thread pool it is processed in the main thread.
 This kinds of falls back to our original processing scheme where every request was just processed in place, but it also allows us to apply backpressure to clients when the server cannot cope with the load.